### PR TITLE
Fix fetches when eval_step's input is a subset of train_step's input

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
This PR also includes MNIST sample using the new forntend as a sample

Example:
`loss, prediction = trainer.train_step(data, target) ` takes `data` and `target` as input and returns `loss` + `prediction`
`prediction = trainer.eval_step(data)` takes `data` and returns `prediction` only

This combination would case an assert to trip incorrectly, as the size of input with fetches can be smaller than input without mismatch  - it is unpredictable actually

### CODE REVIEW

File: samples/python/mnist/mnist_training.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    with torch.no_grad():
+        for data, target in test_loader:
+            # Fetch data
+            data, target = data.to(device), target.to(device)
+            data = data.reshape(data.shape[0], -1)
+
+            # Eval step
+            # Using fetches around without eval_step to not pass 'target' as input
+            trainer._train_step_info.fetches = ['probability']
+            output = F.log_softmax(trainer.eval_step(data), dim=1)
```
Comment (BowenBao) [inline samples/python/mnist/mnist_training.py:70]: what happens if target is still passed in as an argument, would it error out?

Comment (thiagocrepaldi) [reply samples/python/mnist/mnist_training.py:70]: No, it simply doesn't use it. and run eval_step the same way as if target wasnt specified
